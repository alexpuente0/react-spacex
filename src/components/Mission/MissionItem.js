import PropTypes from 'prop-types';
import React from 'react';
import { useDispatch } from 'react-redux';
import { toast } from 'react-toastify';
import { joinMission } from '../../redux/missions/MissionsReducer';

const MissionItem = ({
  id, mission, description, reserved,
}) => {
  const dispatch = useDispatch();
  const handleJoinMission = (e) => {
    const missionId = e.target.className;
    dispatch(joinMission(missionId));
    if (!reserved) {
      toast.success('You just joined a mission successfully!');
    } else {
      toast.success('You just left a mission successfully!');
    }
  };

  return (
    <tr className="mission-container">
      <td className="mission-name">{mission}</td>
      <td>{description}</td>
      <td className="mission">
        {reserved ? (
          <span className="active-member">ACTIVE MEMBER</span>
        ) : (
          <span className="not-member">NOT A MEMBER</span>
        )}
      </td>
      <td className="join-mission">
        <button
          type="button"
          className={id}
          onClick={handleJoinMission}
          style={{
            color: reserved ? 'red' : 'black',
            border: reserved ? '2px solid red' : '2px solid black',
            borderRadius: '0.2rem',
            fontWeight: 'transparent',
            padding: '0.20rem',
            display: 'block',
            margin: 'auto',
            cursor: 'pointer',
          }}
        >
          {reserved ? 'Leave Mission' : 'Join Mission'}
        </button>
      </td>
    </tr>
  );
};

MissionItem.propTypes = {
  id: PropTypes.string.isRequired,
  mission: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  reserved: PropTypes.string.isRequired,
};

export default MissionItem;
