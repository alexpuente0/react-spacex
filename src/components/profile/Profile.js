import React from 'react';
import { useSelector } from 'react-redux';
import MissionProfile from './MissionProfile';

const Profile = () => {
  const rocketState = useSelector((state) => state.rocketsReducer);
  const reservedRockets = rocketState.filter((item) => item.reserved === true);

  const missionState = useSelector((state) => state.missionReducer);
  const reservedMissions = missionState.filter(
    (item) => item.reserved === true,
  );

  return (
    <div className="profile-container">
      <div className="rockets-profile">
        <h2 className="title">My Missions</h2>
        <div className="rockets">
          {reservedMissions.map((item) => (
            <MissionProfile key={item.missionId} name={item.missionName} />
          ))}
        </div>
      </div>
      <div className="rockets-profile">
        <h2 className="title">My Rockets</h2>
        <ul className="rockets">
          {reservedRockets.map((item) => (
            <li className="rockets-li" key={item.id}>
              {item.name}
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};

export default Profile;
