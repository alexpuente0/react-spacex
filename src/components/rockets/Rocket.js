/* eslint-disable */
import PropTypes from "prop-types";
import React from "react";
import { useDispatch } from "react-redux";
import {
	reserveRocket,
	removeReservation,
} from "../../redux/rockets/RocketReducer";
import { toast } from "react-toastify";

const Rocket = ({ id, image, name, description, reserved }) => {
	const dispatch = useDispatch();

	const reserveShip = () => {
		dispatch(reserveRocket(id));
		toast.success("Your Ship was reserved successfully!");
	};

	const cancelReservation = () => {
		dispatch(removeReservation(id));
		toast.success("Your Reservation was canceled successfully!");
	};

	return (
		<div className="rocket-container" id={id}>
			<img src={image} alt="rocket" className="rocket-image" />
			<div className="content">
				<h2 className="name">{name}</h2>
				<p className="description">
					{reserved && <span className="span">Reserved</span>}
					{description}
				</p>
				{reserved && (
					<button
						className="cancel-btn"
						type="button"
						onClick={cancelReservation}
					>
						Cancel Reservation
					</button>
				)}
				{!reserved && (
					<button className="reserve-btn" type="button" onClick={reserveShip}>
						Reserve Rocket
					</button>
				)}
			</div>
		</div>
	);
};

Rocket.propTypes = {
	id: PropTypes.number.isRequired,
	image: PropTypes.string.isRequired,
	name: PropTypes.string.isRequired,
	description: PropTypes.string.isRequired,
	reserved: PropTypes.bool.isRequired,
};

export default Rocket;
