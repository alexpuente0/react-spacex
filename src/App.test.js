/* eslint-disable */
import React from 'react';
import renderer from 'react-test-renderer';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from './redux/configureStore';
import Rocket from './components/rockets/Rocket';
import MissionItem from './components/Mission/MissionItem';
import Profile from './components/profile/Profile';

describe('Pages are rendered correctly', () => {
  it('Renders Missions Page correctly', () => {
    const element = renderer
      .create(
        <Provider store={store}>
          <MissionItem />
        </Provider>,
      )
      .toJSON();
    render(
      <Provider store={store}>
        <MissionItem />
      </Provider>,
    );
    const msRender = document.querySelector('.mission-container');
    expect(msRender).toBeTruthy();
    expect(element).toMatchSnapshot();
  });
  it('Renders Rockets Page correctly', () => {
    const element = renderer
      .create(
        <Provider store={store}>
          <Rocket />
        </Provider>,
      )
      .toJSON();
    render(
      <Provider store={store}>
        <Rocket />
      </Provider>,
    );
    const rkRender = document.querySelector('.rocket-container');
    expect(rkRender).toBeTruthy();
    expect(element).toMatchSnapshot();
  });
  it('Renders My Profile page correctly', () => {
    const element = renderer
      .create(
        <Provider store={store}>
          <Profile />
        </Provider>,
      )
      .toJSON();
    render(
      <Provider store={store}>
        <Profile />
      </Provider>,
    );
    const pfRender = document.querySelector('.profile-container');
    expect(pfRender).toBeTruthy();
    expect(element).toMatchSnapshot();
  });
});
