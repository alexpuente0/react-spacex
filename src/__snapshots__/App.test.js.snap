// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages are rendered correctly Renders Missions Page correctly 1`] = `
<tr
  className="mission-container"
>
  <td
    className="mission-name"
  />
  <td />
  <td
    className="mission"
  >
    <span
      className="not-member"
    >
      NOT A MEMBER
    </span>
  </td>
  <td
    className="join-mission"
  >
    <button
      onClick={[Function]}
      style={
        Object {
          "border": "2px solid black",
          "borderRadius": "0.2rem",
          "color": "black",
          "cursor": "pointer",
          "display": "block",
          "fontWeight": "transparent",
          "margin": "auto",
          "padding": "0.20rem",
        }
      }
      type="button"
    >
      Join Mission
    </button>
  </td>
</tr>
`;

exports[`Pages are rendered correctly Renders My Profile page correctly 1`] = `
<div
  className="profile-container"
>
  <div
    className="rockets-profile"
  >
    <h2
      className="title"
    >
      My Missions
    </h2>
    <div
      className="rockets"
    />
  </div>
  <div
    className="rockets-profile"
  >
    <h2
      className="title"
    >
      My Rockets
    </h2>
    <ul
      className="rockets"
    />
  </div>
</div>
`;

exports[`Pages are rendered correctly Renders Rockets Page correctly 1`] = `
<div
  className="rocket-container"
>
  <img
    alt="rocket"
    className="rocket-image"
  />
  <div
    className="content"
  >
    <h2
      className="name"
    />
    <p
      className="description"
    />
    <button
      className="reserve-btn"
      onClick={[Function]}
      type="button"
    >
      Reserve Rocket
    </button>
  </div>
</div>
`;
